drivers/hmis/app/graphql/mutations/update_external_form_submission.rb

```diff
@@ -8,24 +8,72 @@ module Mutations
     argument :input, Types::HmisSchema::ExternalFormSubmissionInput, required: true
     argument :id, ID, required: true
+    argument :project_id, ID, required: false # not required for backwards compatibility
 
-    field :external_form_submission, Types::HmisSchema::ExternalFormSubmission, null: false
+    field :external_form_submission, Types::HmisSchema::ExternalFormSubmission, null: true
     field :errors, [Types::HmisSchema::ValidationError], null: false, resolver: Resolvers::ValidationErrors
 
-    def resolve(id:, input:)
+    def resolve(...)
+      Hmis::Hud::Base.transaction do
+        _resolve(...)
+      end
+    end
+
+    protected
+
+    EXTRANEOUS_KEYS = ['captcha_score', 'form_definition_id', 'form_content_digest'].freeze
+
+    def _resolve(id:, project_id:, input:)
       record = HmisExternalApis::ExternalForms::FormSubmission.find(id)
       raise 'Access denied' unless allowed?(permissions: [:can_manage_external_form_submissions])
 
       record.assign_attributes(**input.to_params)
 
-      errors = []
-      if record.valid?
-        record.save!
-      else
-        errors = record.errors
-        record = nil
+      errors = HmisErrors::Errors.new
+
+      unless record.valid?
+        errors.add_ar_errors(record.errors&.errors)
+        return { errors: errors }
       end
 
+      if record.status_changed? && record.status == 'reviewed' && !record.spam
+        definition = record.definition
+
+        # Only if there are Client and/or Enrollment fields in the form definition, initialize an enrollment
+        # (which will in turn initialize a Client, inside the form processor).
+        if definition.link_id_item_hash.values.find { |item| ['ENROLLMENT', 'CLIENT'].include?(item.mapping.record_type) }
+          project = Hmis::Hud::Project.find(project_id)
+          record.build_enrollment(project: project, data_source: project.data_source, entry_date: record.created_at)
+          # Assume that required values on Client and Enrollment (e.g. relationship to HoH) are present
+          # and correctly mapped in raw_values. The form processor record validator will fail otherwise.
+        end
+
+        form_processor = record.form_processor || record.build_form_processor(definition: definition)
+
+        form_processor.hud_values = record.raw_data.reject { |key, _| EXTRANEOUS_KEYS.include?(key) }
+        # We skip the form_processor.collect_form_validations step, because the external form has already been
+        # submitted. If it's invalid, there is nothing the user can do about it now.
+
+        # Run to create CDEs, and client/enrollment if applicable
+        form_processor.run!(user: current_user)
+
+        # Collect validations and warnings from AR Validator classes
+        record_validations = form_processor.collect_record_validations(user: current_user)
+        errors.push(*record_validations)
+
+        errors.deduplicate!
+        return { errors: errors } if errors.any?
```

Comment:
I'm not sure if we want this. The user can't do anything to fix the error, it basically means the form is in purgatory until we fix whatever is wrong with it from the backend, or it's deleted.

It looks like `collect_record_validations ` will also run the custom validators (like this [hmis_validate](https://github.com/greenriver/hmis-warehouse/blob/271a749b704e4b200e7fbb1c35f662ed2fcff278/drivers/hmis/app/models/hmis/hud/validators/enrollment_validator.rb#L92-L111)), which are validations that are not strictly needed by the database. I think that's too strict for this scenario.. for example if the submission is resulting in 2 HoHs in one hh, we would completely block them from processing it. I _think_ we want to be more lenient, and allow the submissions to be processed and cleaned up after the fact.

Of course there is another class of validations that will cause `.save!` below to raise an exception. I think it's best to just let those exceptions raise, so that we're notified via Sentry. If we want to error out with a more user-friendly/specific message, we can:
```
  error_out(client.errors.full_messages) unless client.valid?
  error_out(enrollment.errors.full_messages) unless enrollment.valid?
  # proceed to save...
   
    def error_out(msg)
      # error out with user-facing error message
      raise HmisErrors::ApiError.new(msg, display_message: 'Unable to process form submission due to invalid values. Contact your administrator if the problem persists.')
    end
```

That way we can track things in sentry (and mute them if needed), and the user gets something beyond just "an error occurred".

Thoughts? happy to huddle on it!

Reply:
This sounds good to me, so I've updated to use your suggestions. Maybe it would be good to pair on testing this? I think the test case I've come up with will break once we add logic for automatically setting the relationship to HoH, per your comment [here](https://github.com/greenriver/hmis-warehouse/pull/4667#pullrequestreview-2281004233).

Reply:
This should be covered by the test case here: https://github.com/greenriver/hmis-warehouse/pull/4657/files#diff-360e572a5a39d6a50ef9f0365163ca539293ed1cff0617f610d7d316b4d8043cR252-R256